resolver-dns/src/main/java/io/netty/resolver/dns/DnsQueryContext.java

```diff
@@ -148,11 +148,21 @@ public void operationComplete(Future<AddressedEnvelope<DnsResponse, InetSocketAd
                     timeoutFuture.cancel(false);
                 }
 
-                // Remove the id from the manager as soon as the query completes. This may be because of success,
-                // failure or cancellation
-                DnsQueryContext self = queryContextManager.remove(nameServerAddr, id);
-
-                assert self == DnsQueryContext.this : "Removed DnsQueryContext is not the correct instance";
+                if (future.cause() instanceof DnsNameResolverTimeoutException) {
+                    // This query was failed due a timeout. Let's delay the removal of the id to reduce the risk
+                    // of reusing the same id again while the remote nameserver might send the response after the
+                    // timeout. We choose a delay of 20 seconds, after this period all bets are off.
+                    channel.eventLoop().schedule(new Runnable() {
+                        @Override
+                        public void run() {
+                            removeFromContextManager(nameServerAddr);
+                        }
+                    }, 20, TimeUnit.SECONDS);
```

Comment:
should we make `20` configurable? if we start getting timeouts we will pile up used IDs and may exhaust the 16 bits and not be able to send queries until cleanup is triggered. `20` seconds maybe too long depending upon the situation.

Reply:
honestly I feel that if this happens we have bigger problems here...  We could make it a system property but I don't think we should expose it via some API.